### PR TITLE
[3D] On screen navigation

### PR DESCRIPTION
**Pull request by ismailsunni**
## Description
This PR is related to [GSoC 2019 Project: QGIS 3D Improvement](https://github.com/ismailsunni/QGIS/wiki/GSoC-2019---QGIS-3D-Improvement). It implements the on-screen navigation for the 3D map view. The functionalities:
- Move up, down, left, right
- Tilt up and down
- Zoom in and out
- Rotating using compass-like control
- Tool button to show/hide the navigation widget.
- Store its last visibility state to setting

Screenshot:
image

GIF:
3d-navigation

Issue reference: https://github.com/ismailsunni/QGIS/issues/1

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/3d/qgscameracontroller.h [hunk 1]
```diff
@@ -135,6 +135,17 @@ class _3D_EXPORT QgsCameraController : public Qt3DCore::QEntity
     //! Reads camera configuration from the given DOM element
     void readXml( const QDomElement &elem );
 
+    //! Zoom the map by a factor
```
Comment (nyalldawson) [inline src/3d/qgscameracontroller.h]:
```suggestion
    //! Zoom the map by a \a factor
```

File: src/3d/qgscameracontroller.h [hunk 2]
```diff
@@ -135,6 +135,17 @@ class _3D_EXPORT QgsCameraController : public Qt3DCore::QEntity
...
+    void zoom( float factor );
+    //! Tilt up the view by deltaPitch around the view center (camera moves)
```
Comment (nyalldawson) [inline src/3d/qgscameracontroller.h]:
```suggestion
    //! Tilt up the view by \a deltaPitch around the view center (camera moves)
```

Comment (nyalldawson) [reply src/3d/qgscameracontroller.h]: (etc)

File: src/app/3d/qgs3dmapcanvasdockwidget.cpp [hunk 3]
```diff
@@ -147,6 +158,15 @@ void Qgs3DMapCanvasDockWidget::identify()
   mCanvas->setMapTool( action->isChecked() ? mMapToolIdentify : nullptr );
 }
 
+void Qgs3DMapCanvasDockWidget::toggleNavigationWidget()
+{
+  QAction *action = qobject_cast<QAction *>( sender() );
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmapcanvasdockwidget.cpp]: You could avoid the sender check here if you make this a slot with a bool argument and connect it directly to the action::toggled signal. It's a bit nicer then retrieving sender() in a slot.

File: src/app/3d/qgs3dnavigationwidget.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  gridLayout->addWidget( mZoomOutButton, 3, 3 );
+  gridLayout->addWidget( mCompas, 1, 1, 2, 2 );
+  gridLayout->addWidget( mMoveUpButton, 0, 1, 1, 2, Qt::AlignCenter );
+  gridLayout->addWidget( mMoveRightButton, 1, 3, 2, 1, Qt::AlignCenter );
+  gridLayout->addWidget( mMoveDownButton, 3, 1, 1, 2, Qt::AlignCenter );
+  gridLayout->addWidget( mMoveLeftButton, 1, 0, 2, 1, Qt::AlignCenter );
+  gridLayout->setAlignment( Qt::AlignTop );
+}
+
+Qgs3DNavigationWidget::~Qgs3DNavigationWidget()
```
Comment (nyalldawson) [inline src/app/3d/qgs3dnavigationwidget.cpp]:
```suggestion
Qgs3DNavigationWidget::~Qgs3DNavigationWidget() = default
```

(or better yet, remove it totally. It's not needed)

Comment (ismailsunni) [reply src/app/3d/qgs3dnavigationwidget.cpp]: Done

Thanks, @nyalldawson for the review.

File: src/3d/qgscameracontroller.h (hunk 2 above)
Comment (ismailsunni) [reply src/3d/qgscameracontroller.h]: Done for other `\a` also

**Comment by nyalldawson** [comment]
@wonder-sk is this good to merge?

**Comment by wonder-sk** [comment]
Reviewed and tested - working nicely - great job @ismailsunni

**Comment by ismailsunni** [comment]
Thanks @wonder-sk :)

**Pull request merged**